### PR TITLE
[processing] optimise singlepart to multipart algorithm

### PR DESCRIPTION
**Pull request by nirvn**
The PR improves the algorithm by:
- keeping z/m/curved geometries intact
- ~~relying on QgsFeatureRequest to speed things up~~
- reworking the algorithm to avoid looping through features x number of unique values by relying on a dict
- removing the error when no single parts will end up transformed as multi part feature(s) (when occurring, the error was breaking models for justifiable reasons, since the output layer is perfectly fine and usable)

@nyalldawson , thanks for the inspiration.

### CODE REVIEW

File: python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py [hunk 1]
```diff
@@ -75,62 +76,58 @@ def processAlgorithm(self, progress):
...
-            elif wkbType in (QgsWkbTypes.Polygon, QgsWkbTypes.MultiPolygon,
-                             QgsWkbTypes.MultiPolygon25D, QgsWkbTypes.Polygon25D):
-
-                return QgsWkbTypes.MultiPolygon
+
+        for i in unique:
+            if fieldType in [QVariant.Int, QVariant.Double, QVariant.UInt, QVariant.LongLong, QVariant.ULongLong]:
+                expr = '"%s" = %s' % (fieldName, i)
+            elif fieldType == QVariant.String:
+                expr = """"%s" like '%s'""" % (fieldName, i)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py]: Why is this a 'like'?

File: python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py [hunk 2]
```diff
@@ -75,62 +76,58 @@ def processAlgorithm(self, progress):
...
+            for inFeat in features:
+                atMap = inFeat.attributes()
+                idVar = atMap[index]
+                if str(idVar).strip() == str(i).strip():
+                    if first:
+                        attrs = atMap
+                        first = False
+                    inGeom = inFeat.geometry()
+                    vType = inGeom.type()
+                    feature_list = self.extractAsMulti(inGeom)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py]: This part still breaks z/m/curves. I think it should use QgsGeometry::collectGeometry instead

**Review by nyalldawson — COMMENTED** [review]
I'm wondering... If there was many unique values this would still result in lots of feature requests. It might be better to just iterate the whole layer and store the matching features together in a dictionary for that unique field value. And then at the end just collect all the geometry parts and write them out in one go...

**Comment by nirvn** [comment]
@nyalldawson , pushed a commit, which goes at it using the dictionary road, which does prevent unique_values X nb of features. We now have nb of features + unique_values, that's much better.

Am I using the collectGeometry properly?

File: python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py [hunk 3]
```diff
@@ -76,83 +76,34 @@ def processAlgorithm(self, progress):
...
+            inGeom = inFeat.geometry()
+            vType = inGeom.type()
+            collection_geom[key].append(inGeom)
+
+            current += 1
+            progress.setPercentage(int(current * total))
+
+        for key, geoms in collection_geom.items():
+            outFeat.setAttributes(collection_attrs[key])
+            outGeom = QgsGeometry().collectGeometry(geoms)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py]: Just `QgsGeometry.collectGeometry(...)`

File: python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py [hunk 4]
```diff
@@ -76,83 +76,34 @@ def processAlgorithm(self, progress):
...
-            return QgsGeometry().fromMultiPoint(geom_list)
-        elif vType == QgsWkbTypes.LineGeometry:
-            return QgsGeometry().fromMultiPolyline(geom_list)
-        else:
-            return QgsGeometry().fromMultiPolygon(geom_list)
+        for inFeat in features:
+            atMap = inFeat.attributes()
+            idVar = atMap[index]
+            key = str(idVar).strip()
+            if not key in collection_geom.keys():
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py]: This can be optimised to `if not key in collection _geom:`

**Review by nyalldawson — COMMENTED** [review]
Looks much better!

**Comment by nirvn** [comment]
@nyalldawson , commit updated with your comments addressed. Thanks for the review.

File: python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py (hunk 3 above)
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py]: You've still got the extra () in there. It's a static method, so isn't called on an object

Comment (nirvn) [inline python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py]: @nyalldawson , oups, fixed now.

**Comment by ahuarte47** [comment]
Hi @nirvn and @nyalldawson, I am studing this optimised processing algorithm and I see the feature collection is fetched twice, [here](https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py#L82) and [here](https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/qgis/SinglePartsToMultiparts.py#L86). 

Many other algorithms iterate features using:

`features = vector.features(layer)`
`total = 100.0 / len(features)`
`for current, f in enumerate(features):`
`   bla bla`

Is it a bug or a performance bottlenek for large data stores? or am I wrong?
My apologies if I am wrong

**Comment by nyalldawson** [comment]
Oops - definitely a bug!

**Comment by ahuarte47** [comment]
Could I propose this fix?

https://github.com/qgis/QGIS/pull/3746

**Pull request merged**